python/cudf/cudf/core/column/categorical.py

```diff
@@ -281,9 +281,35 @@ def add_categories(self, new_categories, **kwargs):
         Categories (5, int64): [1, 2, 0, 3, 4]
         """
         new_categories = column.as_column(new_categories)
-        new_categories = self._column.categories.append(new_categories)
+        old_categories = self._column.categories
+
+        if (
+            old_categories.dtype == "object"
+            and new_categories.dtype != "object"
+        ) or (
+            old_categories.dtype != "object"
+            and new_categories.dtype == "object"
+        ):
+            raise TypeError(
+                "cudf doesnot support mixed types, please type-cast \
+                    new_categories to the same type as existing categories."
+            )
+        common_dtype = np.find_common_type(
+            [old_categories.dtype, new_categories.dtype], []
+        )
+
+        if new_categories.dtype != common_dtype:
+            new_categories = new_categories.astype(common_dtype)
+        elif old_categories.dtype != common_dtype:
+            old_categories = old_categories.astype(common_dtype)
```

Comment:
Can just use zero-copy typecasting here:
```suggestion
        new_categories = new_categories.astype(common_dtype, copy=False)
        old_categories = old_categories.astype(common_dtype, copy=False)
```